### PR TITLE
bug: cmdline args need jsonification; tty support needs adjustment

### PR DESCRIPTION
**Pull request by joesuf4**
- switch shell to bash so we can use arrays and bulk element transforms
- use [ -t 0 ] to baseline whether we need terminal support for the invocation

### CODE REVIEW

**Comment by joesuf4** [comment]
The motivation for this patch is that I want to use node-shell in CLI pipelines, so I am not interested in the default tty login shell API.

I don't want "stuff" printed to stdout unless I'm expecting it from my CLI invocation, so I redirect that output to stderr.  Also I have embedded newlines and double-quote characters in some of my CLI tooling that wraps node-shell, so I need better CLI->json-string translation than before I needed now.

**Comment by kvaps** [comment]
Hi thank you for your PR, I'll review it in a while 👍

File: kubectl-node_shell [hunk 1]
```diff
@@ -68,16 +68,18 @@ done
...
 else
-  cmd="$cmd\"bash\", \"-l\" ]"
+  cmd+=(bash -l)
 fi
+# translate embedded single-quotes to double-quotes, so the following line will work
+cmd=( "${cmd[@]//\'/\"}" )
+
+# jsonify(as an array) the argument list (mainly from the command line)
+entrypoint="$(echo "['${cmd[@]/%/\', \'}']" | sed -e "s/' /'/g" \
+                   -e "s/, '']\$/]/" -Ee "s/([\"\\])/\\\\\1/g" -e 's/\\\\n/\\n/g' | tr \' \")"
```
Comment (kvaps) [inline kubectl-node_shell:82]:
```suggestion
```

File: kubectl-node_shell [hunk 2]
```diff
@@ -107,7 +109,7 @@ overrides="$(
         "stdin": true,
         "stdinOnce": true,
         "tty": $tty,
-        "command": $cmd
+        "command": $entrypoint
```
Comment (kvaps) [inline kubectl-node_shell:112]:
```suggestion
        "command": $cmd
```

File: kubectl-node_shell [hunk 3]
```diff
@@ -131,7 +133,7 @@ if [ "$m" -lt 18 ]; then
   generator="--generator=run-pod/v1"
 fi
 
-trap "EC=\$?; $kubectl delete pod --wait=false $pod 2>/dev/null || true; exit \$EC" EXIT INT TERM
+trap "EC=\$?; $kubectl delete pod --wait=false $pod >&2 || true; exit \$EC" EXIT INT TERM
 
-echo "spawning \"$pod\" on \"$node\""
-$kubectl run --image "$image" --restart=Never --overrides="$overrides" -ti "$pod" $generator
+echo "spawning \"$pod\" on \"$node\"" >&2
+$kubectl run --image "$image" --restart=Never --overrides="$overrides"  $([ -t 0 ] && echo -t) -i "$pod" $generator
```
Comment (kvaps) [inline kubectl-node_shell:139]:
```suggestion
$kubectl run --image "$image" --restart=Never --overrides="$overrides"  $([ "$tty" = true ] && echo -t) -i "$pod" $generator
```

**Review by kvaps — COMMENTED** [review]
Hi @joesuf4, sorry for late reply.
Your PR adds two new dependencies `bash` and `sed`.

I agree that using bash arrays makes it more convenient, but the original goal for this plugin was to keep it simplest and compatible with all possible architectures and any shell.

I believe that there is no big reason to upgrade shell version. Please take a look on these changes, I think I was able to implement your changes compatible with standard bourne shell

File: kubectl-node_shell [hunk 4]
```diff
@@ -68,16 +68,18 @@ done
...
-while [ $# -gt 0 ]; do
-  cmd="$cmd\"$1\", "
-  custom=true
-  shift
-done
-if [ "$custom" = true ]; then
-  cmd="$(echo "$cmd" | awk '{sub(/..$/,"",$0); print $0}') ]"
+if [ $# -gt 0 ]; then
+  cmd+=( "${@//
+/\\n}" )
```
Comment (kvaps) [inline kubectl-node_shell:73]:
```suggestion
  while [ $# -gt 0 ]; do
    cmd="$cmd, \"$1\""
    shift
  done
  cmd="$cmd ]"
```

File: kubectl-node_shell (hunk 1 above)
Comment (joesuf4) [reply kubectl-node_shell:82]: You just nuked 95% of the point of the patch,  which is to do impedance matching between nontrivial command line arguments to kubectl node-shell, and the JSON array you must serialize them with. Here is a trivial example to test (in a bash shell CLI):
kubectl node-shell $node — sh -c “

echo oops”



> Hi @joesuf4, sorry for late reply.
> Your PR adds two new dependencies `bash` and `sed`.
> 
> I agree that using bash arrays makes it more convenient, but the original goal for this plugin was to keep it simplest and compatible with all possible architectures and any shell.
> 
> I believe that there is no reasons to upgrade shell version. Please take a look on these changes, I think I was able to make your changes compatible with standard bourne shell

This makes no sense.  Any user with a kubectl install also has a bash + see install, so you are just complicating your life for no gain by using a primitive shell to solve a complex data serialization problem.

Comment (joesuf4) [reply kubectl-node_shell:82]: *bash + sed, stupid autocorrect.

File: kubectl-node_shell [hunk 5]
```diff
@@ -1,14 +1,14 @@
...
 kubectl=kubectl
 generator=""
 node=""
 nodefaultctx=0
 nodefaultns=0
-cmd='[ "nsenter", "--target", "1", "--mount", "--uts", "--ipc", "--net", "--pid", "--", '
+cmd=(nsenter --target 1 --mount --uts --ipc --net --pid --)
 custom=false
-if ! [ -p /dev/stdin ] && ! [ -p /dev/stdout ]; then
+if [ -t 0 ]; then
```
[the remark this answers is not in the document]
Comment (joesuf4) [reply kubectl-node_shell:11]: This is incorrect logic for dealing with tty settings in a kubectl environment. The only thing that matters to kubectl is whether or not stdin is a tty.  The device type of stdout is irrelevant; furthermore there are other device types than fifos and tty’s for a Unix file descriptor, so why beat around the bush by not using [ -t 0 ]?

File: kubectl-node_shell (hunk 1 above)
Comment (kvaps) [reply kubectl-node_shell:82]: > You just nuked 95% of the point of the patch

Sorry 😊

> Here is a trivial example to test (in a bash shell CLI):


OK, got it! But how about quotes?

```bash
kubectl node-shell $node -- sh -c '

echo o"ops'
```

**Review by joesuf4 — COMMENTED** [review]
Are we agreed on this MR?

**Comment by joesuf4** [comment]
Ok.

**Comment by joesuf4** [comment]
Fighting with GitHub MR ui.  Bear with me.

**Comment by kvaps** [comment]
Merged in `v1.4.0`.
Many thanks for your contribution!

**Pull request merged**